access-common/src/main/java/edu/unc/lib/dl/ui/service/FedoraContentService.java

```diff
@@ -15,235 +15,98 @@
+import static org.apache.http.HttpHeaders.CONTENT_LENGTH;
+import static org.apache.http.HttpHeaders.CONTENT_TYPE;
+
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.List;
+import java.io.OutputStream;
 
 import javax.servlet.http.HttpServletResponse;
 
+import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.springframework.beans.factory.annotation.Autowired;
 
+import edu.unc.lib.dl.acl.service.AccessControlService;
 import edu.unc.lib.dl.acl.util.AccessGroupSet;
-import edu.unc.lib.dl.acl.util.GroupsThreadStore;
-import edu.unc.lib.dl.fedora.AccessClient;
-import edu.unc.lib.dl.fedora.AuthorizationException;
-import edu.unc.lib.dl.fedora.FedoraException;
-import edu.unc.lib.dl.search.solr.model.BriefObjectMetadataBean;
-import edu.unc.lib.dl.search.solr.model.Datastream;
-import edu.unc.lib.dl.search.solr.model.SimpleIdRequest;
-import edu.unc.lib.dl.search.solr.util.SearchFieldKeys;
-import edu.unc.lib.dl.search.solr.util.SearchSettings;
-import edu.unc.lib.dl.ui.exception.InvalidRecordRequestException;
-import edu.unc.lib.dl.ui.exception.ResourceNotFoundException;
-import edu.unc.lib.dl.ui.util.AnalyticsTrackerUtil;
-import edu.unc.lib.dl.ui.util.AnalyticsTrackerUtil.AnalyticsUserData;
-import edu.unc.lib.dl.util.ContentModelHelper;
-import edu.unc.lib.dl.util.ContentModelHelper.DatastreamCategory;
+import edu.unc.lib.dl.fcrepo4.BinaryObject;
+import edu.unc.lib.dl.fcrepo4.FileObject;
+import edu.unc.lib.dl.fcrepo4.RepositoryObjectLoader;
+import edu.unc.lib.dl.fedora.PID;
 
 /**
- * Connects to and streams datastreams from Fedora.
+ * Streams binary content from repository objects.
  *
  * @author bbpennel
  */
 public class FedoraContentService {
     private static final Logger LOG = LoggerFactory.getLogger(FedoraContentService.class);
 
-    private AccessClient accessClient;
-
-    private String fedoraHost;
-
-    @Autowired
-    private SearchSettings searchSettings;
-    @Autowired
-    protected SolrQueryLayerService queryLayer;
-
-    @Autowired(required = false)
-    protected AnalyticsTrackerUtil analyticsTracker;
-
-    private final int numberOfRetries = 1;
+    public static final String CONTENT_DISPOSITION = "Content-Disposition";
+
+    private AccessControlService accessControlService;
+
+    private RepositoryObjectLoader repositoryObjectLoader;
+
+    /**
+     * Set content headers and stream the binary content of the specified
+     * datastream from the object identified by pid.
+     *
+     * @param pid pid of object containing datastream
+     * @param dsName name of datastream being requested. If null, then original
+     *            file assume.
+     * @param principals principals of requesting client
+     * @param asAttachment if true, then content-disposition header will specify
+     *            as "attachment" instead of "inline"
+     * @param response response content and headers will be added to.
+     * @throws IOException
```

Comment:
perhaps consider adding a description about how/when this exception gets thrown?

Reply:
I will add this but we have often not included descriptions for @throws, are you interested in changing that going forward? We do most likely need a more intentional approach to what we throw and don't throw. My tendency is to add descriptions if it seems like the throw is part of the api that the code calling it will need to know to expect and recover from. Perhaps that means that in other cases we should be recasting to a runtime exception instead.

Reply:
i do think descriptions help to understand what to expect from a method, but also agree with you it's most important in the case of exceptions when the throw is part of the api the calling code will need to know to expect and recover from. i would say if the calling code doesn't need to know about and/or be able to recover from a certain exception, maybe it shouldn't be in the javadoc. and 👍  to the idea of a more intentional approach!